PowerEditor/src/WinControls/ProjectPanel/Directory.cpp

```diff
@@ -0,0 +1,454 @@
+					if (!containsData(childPath))
+					{
+						_invisibleDirs.push_back(file);
+						if (!FindNextFile(hFind, &fd))
+							break;
+						continue;
+					}
+				}
+				_dirs.insert(file);
+			}
+		}
+		else
+		{
+			if (readFiles)
+				_files.insert(file);
+		}
+
+		if (!FindNextFile(hFind, &fd))
+			break;
+	}
+
+	if (hFind != INVALID_HANDLE_VALUE)
+		FindClose(hFind);
+
+
+}
+
+
+bool Directory::containsDataChanged() const
+{
+	for (auto it=_dirs.begin(); it != _dirs.end(); ++it)
+	{
+		generic_string childPath = _path + TEXT("\\") + *it;
+		if (!containsData(childPath))
+			return true;
+	}
+	for (auto it=_invisibleDirs.begin(); it != _invisibleDirs.end(); ++it)
+	{
+		generic_string childPath = _path + TEXT("\\") + *it;
+		if (containsData(childPath))
+			return true;
+	}
+	return false;
+}
+
+// containsData() checks, if a directory or its subdirs contain any data which match the filters.
+bool Directory::containsData(const generic_string& path) const
+{
+	if (_filters.empty())
+		return containsData(path, TEXT("*.*"));
+
+	for (size_t i=0; i<_filters.size(); ++i)
+	{
+		if (containsData(path, _filters[i]))
+			return true;
+	}
+	return false;
+}
+
+// containsData() recursive call
+bool Directory::containsData(const generic_string& path, const generic_string& filter) const
+{
+	generic_string searchPath = path + TEXT("\\") + filter;
+	generic_string childPath;
+	generic_string file;
+
+
+	// first step: check only files. If one is found, it contains data.
+	WIN32_FIND_DATA fd;
+	HANDLE hFind = FindFirstFileEx(searchPath.c_str(), FindExInfoBasic, &fd, FindExSearchNameMatch, NULL, 0);
+
+	if (hFind == INVALID_HANDLE_VALUE)
+		return false;
+
+	while (hFind != INVALID_HANDLE_VALUE)
+	{
+
+		// a directory was found... skip it in this pass
+		if ((fd.dwFileAttributes & FILE_ATTRIBUTE_DIRECTORY) != 0)
+		{
+			if (!FindNextFile(hFind, &fd))
+				return false;
+			continue;
+		}
+		else
+		{
+			// data was found - return true
+			FindClose(hFind);
+			return true;
+		}
+
+	}
+
+	if (hFind != INVALID_HANDLE_VALUE)
+		FindClose(hFind);
+
+	// no file was found in this directory. Now check the sub directories.
+	hFind = FindFirstFileEx(searchPath.c_str(), FindExInfoBasic, &fd, FindExSearchLimitToDirectories, NULL, 0);
+
+	if (hFind == INVALID_HANDLE_VALUE)
+		return false;
+
+	while (hFind != INVALID_HANDLE_VALUE)
+	{
+
+		// a directory was found
+		if ((fd.dwFileAttributes & FILE_ATTRIBUTE_DIRECTORY) != 0)
+		{
+			file = fd.cFileName;
+			// skip . and ..
+			if (file == TEXT(".") || file == TEXT(".."))
+			{
+				if (!FindNextFile(hFind, &fd))
+					break;
+				continue;
+			}
+
+			// check if this child directory contains data; if this is the case, return true
+			childPath = path + TEXT("\\") + file;
+			if (containsData(childPath))
+			{
+				FindClose(hFind);
+				return true;
+			}
+		}
+		else
+		{
+			// data was found - return true. This should never happen, because the files were already checked. Just to be safe.
```

Comment:
Oops... no. The assumption that this can not happen is wrong, because a file could be created between the 1st and 2nd iteration. I'll change the comment.